### PR TITLE
WIP: Clean version of semi-supervised PR

### PR DESCRIPTION
**Pull request by vimalmanohar**

### CODE REVIEW

**Comment by vimalmanohar** [comment]
@pegahgh 
Can you please review this pull request?

**Comment by pegahgh** [comment]
sure!

**Comment by vimalmanohar** [comment]
@hhadian Could you please review this PR?

**Comment by hhadian** [comment]
Sure, I will do it.

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_100k_a.sh [hunk 1]
```diff
@@ -0,0 +1,199 @@
+#!/bin/bash
+
+# Copyright 2017  Vimal Manohar
+# Apache 2.0
+
+# This is fisher chain recipe for training a model on a subset of around 100 hours.
```
Comment (pegahgh) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_100k_a.sh]: Does this script use 100hrs supervised training data? add better description e.g. this script uses 100hrs supervised data

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_100k_a.sh [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+get_egs_stage=-10
+decode_iter=
+train_set=train_sup
+tree_affix=bi_a
+nnet3_affix=
+chain_affix=
+exp=exp/semisup_100k
+gmm=tri4a
+xent_regularize=0.1
+hidden_dim=725
```
Comment (pegahgh) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_100k_a.sh]: Isn't it large? we use 625 for 300hrs swbd data.

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_100k_a.sh [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+# training options
+num_epochs=4
+remove_egs=false
+common_egs_dir=
+minibatch_size=128
```
Comment (pegahgh) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_100k_a.sh]: fix it in the script.

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_100k_a_oracle.sh [hunk 4]
```diff
@@ -0,0 +1,207 @@
+#!/bin/bash
+set -e
+
+# This is an oracle experiment using oracle transcription of 250 hours of 
+# unsupervised data, along with 100 hours of supervised data.
```
Comment (pegahgh) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_100k_a_oracle.sh]: I think you can easily use run_tdnn_100k_a.sh with new combined dataset, I am not sure why do you need two separate scripts?

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_i.sh [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+train_set=train_sup15k
+unsup_train_set=unsup_250k
+semisup_train_set=semisup15k_250k
+tree_affix=bi_i
+nnet3_affix=_semi15k_250k
+chain_affix=_semi15k_250k
+exp=exp/semisup_15k
+gmm=tri3
+xent_regularize=0.1
+hidden_dim=500
```
Comment (pegahgh) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_i.sh]: Did you try to reduce it to smaller size or reducing number of layers?

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_j.sh [hunk 6]
```diff
@@ -0,0 +1,203 @@
+#!/bin/bash
+
+# Copyright 2017  Vimal Manohar
+# Apache 2.0
+
+set -e
+
+# This is fisher chain recipe for training a model on a subset of around 15 hours.
+# This is similar to _d, but uses a speed-perturbed data for tree building
```
Comment (pegahgh) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_j.sh]: Why do you keep two scripts _d and _j?  It is better to keep the better one.

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+  echo "$0: chain training on the supervised subset data/${supervised_set}"
+  local/chain/run_tdnn_15k_j.sh $train_supervised_opts \
+                          --train-set $supervised_set \
+                          --unsup-train-set $unsupervised_set \
+                          --semisup-train-set $semisup_train_set \
+                          --nnet3-affix "$nnet3_affix" --tdnn-affix "$tdnn_affix" --exp $exp || exit 1
+fi
+
+lang=data/lang_chain
+unsup_decode_lang=data/lang_poco_test_ex250k
```
Comment (pegahgh) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh]: add brief description for rescore LM and decode LM! Did you generate poco LM using 15 hrs training data?

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+lattice_prune_beam=4.0  # If supplied will prune the lattices prior to getting egs for unsupervised data
+tolerance=1
+phone_insertion_penalty=
+
+rescore_unsup_lattices=true
+unsup_rescoring_affix=big
+
+# Semi-supervised options
+comb_affix=comb1am  # affix for new chain-model directory trained on the combined supervised+unsupervised subsets
+supervision_weights=1.0,1.0
```
Comment (pegahgh) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh]: add description

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+lm_weights=5,2
+
+sup_egs_dir=   # Supply this to skip supervised egs creation
+unsup_egs_dir=  # Supply this to skip unsupervised egs creation
+unsup_egs_opts=  # Extra options to pass to unsupervised egs creation
+
+# Neural network opts
+apply_deriv_weights=true
+xent_regularize=0.1
+hidden_dim=725
```
Comment (pegahgh) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh]: Did you try to tune it? my guess is that it is large hidden dim.

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+minibatch_size="150=128/300=64"
```
Comment (pegahgh) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh]: Is it better than using minibatch_size=150,300?

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+  # has the effect of regularizing the hidden parts of the model.  we use
+  # 0.5 / args.xent_regularize as the learning rate factor- the factor of
+  # 0.5 / args.xent_regularize is suitable as it means the xent
+  # final-layer learns at a rate independent of the regularization
+  # constant; and the 0.5 was tuned so as to make the relative progress
+  # similar in the xent and regular final layers.
+  relu-batchnorm-layer name=prefinal-xent input=tdnn6 dim=$hidden_dim target-rms=0.5
+  output-layer name=output-xent dim=$num_targets learning-rate-factor=$learning_rate_factor max-change=1.5
+  
+  output name=output-0 input=output.affine skip-in-init=true
```
Comment (pegahgh) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh]: Why do you have two separate output nodes? Do you use weighted training? supervision weights were the same in the script.

Comment (pegahgh) [reply egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh]: what is skip-in-init option? You can add a single line comment about output in config file.

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+               --left-tolerance $tolerance --right-tolerance $tolerance \
+               --left-context $left_context --right-context $right_context \
+               --left-context-initial $left_context_initial --right-context-final $right_context_final \
+               --frames-per-eg $unsup_frames_per_eg --frames-per-iter 1500000 \
+               --frame-subsampling-factor $frame_subsampling_factor \
+               --cmvn-opts "$cmvn_opts" --lattice-lm-scale $lattice_lm_scale \
+               --lattice-prune-beam "$lattice_prune_beam" \
+               --phone-insertion-penalty "$phone_insertion_penalty" \
+               --deriv-weights-scp $chaindir/best_path_${unsupervised_set}${decode_affix}/weights.scp \
+               --online-ivector-dir $exp/nnet3${nnet3_affix}/ivectors_${semisup_train_set}_sp_hires \
```
Comment (pegahgh) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh]: Do you use supervised data for ivector training?

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh (hunk 11 above)
Comment (vimalmanohar) [reply egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh]: skip-in-init is added to prevent output line (trivial output layer) from being printed in init.config.
Do you know why the trivial output layers are needed in init.config?

Comment (pegahgh) [reply egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_semisupervised_conf_am.sh]: I think we just use init.config for training lda matrix and we don't need to add other outputs. Probably, we can modify the xconfig to not print trivial outputs in init.config (we just need to print output-node name=output).

File: egs/fisher_english/s5/local/fisher_train_lms.sh [hunk 13]
```diff
@@ -70,6 +77,8 @@ cat $cleantext | awk -v wmap=$dir/word_map 'BEGIN{while((getline<wmap)>0)map[$1]
 
 train_lm.sh --arpa --lmtype 3gram-mincount $dir || exit 1;
 
+train_lm.sh --arpa --lmtype 4gram-mincount $dir || exit 1;
```
Comment (hhadian) [inline egs/fisher_english/s5/local/fisher_train_lms.sh]: If you are using pocolm, it might be better to leave this script unchanged (to make the PR smaller)

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_100k_a_oracle.sh (hunk 4 above)
Comment (hhadian) [reply egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_100k_a_oracle.sh]: I agree. In the inital PR, there was only one TDNN recipe which was
called with different training data sets during semi-supervised training.
Separating the scripts might be clearer but it will add too many very similar scripts.

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_i.sh (hunk 5 above)
Comment (hhadian) [reply egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_i.sh]: I guess I tried smaller sizes but it did not help much.

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_i.sh [hunk 14]
```diff
@@ -0,0 +1,201 @@
+#!/bin/bash
```
Comment (hhadian) [inline egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_i.sh]: I guess it would be nicer to keep only one version of everything (even though it's in tuning) so no _i, _a, etc. That's because there are too many files in this PR.

File: src/online2bin/extend-wav-with-silence.cc [hunk 15]
```diff
@@ -76,16 +76,47 @@ int main(int argc, char *argv[]) {
       exit(1);
     }
 
-    std::string wav_rspecifier = po.GetArg(1);
-    std::string wav_wspecifier = po.GetArg(2);
+    if (ClassifyRspecifier(po.GetArg(1), NULL, NULL) != kNoRspecifier) {
```
Comment (hhadian) [inline src/online2bin/extend-wav-with-silence.cc]: This change was for perturb_to_allowed_lengths.py. I guess you are not using
that script (i.e. non-split training) so it might be better to leave this file unchanged.

File: src/nnet3/nnet-chain-example.cc [hunk 16]
```diff
@@ -51,8 +51,11 @@ void NnetChainSupervision::Read(std::istream &is, bool binary) {
   ReadToken(is, binary, &token);
   // in the future this back-compatibility code can be reworked.
   if (token != "</NnetChainSup>") {
-    KALDI_ASSERT(token == "<DW>");
-    ReadVectorAsChar(is, binary, &deriv_weights);
+    KALDI_ASSERT(token == "<DW>" || token == "<DW2>");
+    if (token == "<DW>")
+      ReadVectorAsChar(is, binary, &deriv_weights);
+    else
+      deriv_weights.Read(is, binary);
```
Comment (hhadian) [inline src/nnet3/nnet-chain-example.cc:58]: how is `<DW2>` diffrent from `<DW>`?

**Review by hhadian — COMMENTED** [review]
I was under the impression that this PR could be wrapped up in 20-30 changed files (new or modified)
Not sure but I feel 70 changed files is too many.

**Comment by hhadian** [comment]
BTW, I noticed your `master` is not up-to-date with upstream kaldi. 
Some of the changes in this PR might be already in kaldi master.

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_i.sh (hunk 5 above)
Comment (vimalmanohar) [reply egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_i.sh]: I think the Fisher dev and test are more similar to training compared to eval2000 and swbd. So it might be ok for a larger network.

File: egs/fisher_english/s5/local/fisher_train_lms.sh (hunk 13 above)
Comment (vimalmanohar) [reply egs/fisher_english/s5/local/fisher_train_lms.sh]: Ok makes sense.

File: src/nnet3/nnet-chain-example.cc (hunk 16 above)
Comment (vimalmanohar) [reply src/nnet3/nnet-chain-example.cc:58]: DW reads only 0 and 1. DW2 reads and writes as float, which is needed for this.

File: egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_j.sh (hunk 6 above)
Comment (vimalmanohar) [reply egs/fisher_english/s5/local/semisup/chain/tuning/run_tdnn_15k_j.sh]: I removed _d. But there are two versions, one with UNK model using phone LM and the other without.
